### DISCUSSION TITLE
[Extending proposal] MetaStream and BLAST

### DISCUSSION THREAD

**Discussion by maffei2443** [online-ml/river#945] [Ideas]
Hello, River Team.

I've developed [a little library](https://github.com/maffei2443/metastream) which implements two (or three) frameworks for algorithm selection, [MetaStream](https://www.sciencedirect.com/science/article/pii/S0925231213007819) and [BLAST](https://www.researchgate.net/publication/289252355_Having_a_Blast_Meta-Learning_and_Heterogeneous_Ensembles_for_Data_Streams).  
I would like to adapt the code, as well documenting it, so that it could be integrated into River.  
Is that possible?  Which patterns and "best practices" am I supposed to follow?  

I know the current code needs a lot of improvement for people other than me using it of course, but I think it may fit the library purpose.

**Comment by MaxHalford** [comment]
Hello 👋

You linked the `statsmodels` library, but I'm guessing you referring to [this](https://github.com/maffei2443/metastream) repo of yours?

Anyway, we're definitely open to contributions :). Here is some advice:

- Start with a single method. You can always port the other one(s) later.
- [`CONTRIBUTING.md`](https://github.com/online-ml/river/blob/main/CONTRIBUTING.md) has instructions on our development workflow. Please let us know if it's not clear enough :)
- Reading some River code is the best way to understand how we do things.
- In your case, the methods you suggest are ensemble methods. Therefore I advise you get familiar with the `ensemble` module.
- The reason why they're ensemble methods is that all the base models are updated, and the output are getting aggregated in some way. This is different to the `model_selection` module, where the outputs are not aggregated, and instead a subset of models is updated and the best model is used to make predictions. Of course ensemble and model selection methods can be compared in terms of performance.

I hope that helps! We're all happy to answer questions you may have.
